packages/core/src/lib/utils/providers.ts

```diff
@@ -9,65 +10,85 @@ import type {
   ProfileCallback,
   Provider,
 } from "../../providers/index.js"
-import type { InternalProvider, Profile } from "../../types.js"
+import type { Awaitable, InternalProvider, Profile } from "../../types.js"
 import type { AuthConfig } from "../../index.js"
 
 /**
  * Adds `signinUrl` and `callbackUrl` to each provider
  * and deep merge user-defined options.
  */
-export default function parseProviders(params: {
+export default async function parseProviders(params: {
   providers: Provider[]
   url: URL
   providerId?: string
   options: AuthConfig
-}): {
+}): Promise<{
   providers: InternalProvider[]
   provider?: InternalProvider
-} {
+}> {
   const { providerId, options } = params
   const url = new URL(options.basePath ?? "/auth", params.url.origin)
 
-  const providers = params.providers.map((p) => {
-    const provider = typeof p === "function" ? p() : p
-    const { options: userOptions, ...defaults } = provider
+  const providers = await Promise.all(
+    params.providers.map<Awaitable<InternalProvider>>(async (p) => {
+      const provider = typeof p === "function" ? p() : p
+      const { options: userOptions, ...defaults } = provider
 
-    const id = (userOptions?.id ?? defaults.id) as string
-    // TODO: Support if properties have different types, e.g. authorization: string or object
-    const merged = merge(defaults, userOptions, {
-      signinUrl: `${url}/signin/${id}`,
-      callbackUrl: `${url}/callback/${id}`,
-    })
+      const id = (userOptions?.id ?? defaults.id) as string
+      // TODO: Support if properties have different types, e.g. authorization: string or object
+      const merged = merge(defaults, userOptions, {
+        signinUrl: `${url}/signin/${id}`,
+        callbackUrl: `${url}/callback/${id}`,
+      })
 
-    if (provider.type === "oauth" || provider.type === "oidc") {
-      merged.redirectProxyUrl ??= options.redirectProxyUrl
-      return normalizeOAuth(merged)
-    }
+      if (provider.type === "oauth" || provider.type === "oidc") {
+        merged.redirectProxyUrl ??= options.redirectProxyUrl
+        return normalizeOAuth(merged)
+      }
 
-    return merged
-  })
+      return merged
+    })
+  )
 
   return {
     providers,
     provider: providers.find(({ id }) => id === providerId),
   }
 }
 
-// TODO: Also add discovery here, if some endpoints/config are missing.
-// We should return both a client and authorization server config.
-function normalizeOAuth(
-  c: OAuthConfig<any> | OAuthUserConfig<any>
-): OAuthConfigInternal<any> | {} {
-  if (c.issuer) c.wellKnown ??= `${c.issuer}/.well-known/openid-configuration`
+async function normalizeOAuth(c: OAuthConfig<any> | OAuthUserConfig<any>) {
+  let authorization: OAuthConfigInternal<any>["authorization"] | undefined
+  if (c.issuer) {
+    const issuer = new URL(c.issuer)
+    try {
+      const discoveryResponse = await o.discoveryRequest(issuer)
```

Comment:
we can cache this in the module scope to potentially hit it less

Reply:
Not sure what do you mean by "module scope" but I implemented the caching mechanism with the cookies. My worry is that this could be considered bloat :/

Reply:
another solution is to make authorization become a lazy function for oidc providers. Technically, it's not much different with the original code, just move the discovery function call to providers, and parseProviders won't need to be async either. This is the first solution I've thought of.